File: core/composition/Composition.cpp
/* Copyright (c) 2011, NEC Europe Ltd, Consorzio Nazionale 
 * Interuniversitario per le Telecomunicazioni, Institut 
 * Telecom/Telecom Bretagne, ETH Zürich, INVEA-TECH a.s. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are met:
 *    * Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    * Redistributions in binary form must reproduce the above copyright
 *      notice, this list of conditions and the following disclaimer in the
 *      documentation and/or other materials provided with the distribution.
 *    * Neither the names of NEC Europe Ltd, Consorzio Nazionale 
 *      Interuniversitario per le Telecomunicazioni, Institut Telecom/Telecom 
 *      Bretagne, ETH Zürich, INVEA-TECH a.s. nor the names of its contributors 
 *      may be used to endorse or promote products derived from this software 
 *      without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR 
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT 
 * HOLDERBE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR 
 * PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER 
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE
 */




#include <Composition.hpp>
#include <CompositionManager.hpp>

using namespace blockmon;


OutGate& Composition::find_output_gate( const std::string& from_block,const std::string& from_gate)
{
    auto map_it = m_map.find(from_block);
    if (map_it == m_map.end())
        throw std::runtime_error(std::string("Cannot create connection from block '").append(
                                 std::string(from_block)).append("': no such block."));
    std::shared_ptr<Block> src_block = map_it->second;
    return src_block->get_output_gate(from_gate); //throws if gate is missing
}


InGate& Composition::find_input_gate( const std::string& to_block,const std::string& to_gate)
{
    auto map_it = m_map.find(to_block);
    if (map_it == m_map.end())
        throw std::runtime_error(std::string("Cannot create connection to block '").append(
                                 std::string(to_block)).append("': no such block."));
    std::shared_ptr<Block> dst_block = map_it->second;
    return dst_block->get_input_gate(to_gate); //throws if gate is missing
}




void Composition::create_link(const std::string& from_block,const std::string& from_gate ,const std::string& to_block,const std::string& to_gate)
{
    OutGate& from = find_output_gate(from_block,from_gate);
    InGate& to = find_input_gate(to_block,to_gate);
    to.connect(from);
    from.connect(to);

}
void Composition::delete_link(const std::string& from_block,const std::string& from_gate ,const std::string& to_block,const std::string& to_gate)
{
    OutGate& from = find_output_gate(from_block,from_gate);
    InGate& to = find_input_gate(to_block,to_gate);
    to.disconnect(&from);
    from.disconnect(&to);
}

void Composition::create_block_from_parsed(const std::string& type, 
                                  const std::string& name, 
                                  invocation_type invocation, 
                                  const pugi::xml_node & blocknode)
{
    std::shared_ptr<Block> sp = BlockFactory::instantiate(type, name, invocation);
    if (!sp) {
        throw std::runtime_error(std::string(type).append(": block type not supported"));
    }
    sp->configure(blocknode);
    m_map[name] = std::move(sp);
}

/*
 * Invoke initialize method on each configured block
 */

void Composition::initialize()
{
    auto it = m_map.begin(),
         it_e = m_map.end();
    
    for(; it != it_e; ++it)
    {
        std::cout << "Composition: Initializing " << it->first << " block" << std::endl;
        it->second->initialize();
    }
}


void Composition::install(const pugi::xml_node& config)
{
    for (pugi::xml_node blocknode=config.child("block"); blocknode; blocknode=blocknode.next_sibling("block"))
    {
        std::string name = blocknode.attribute("id").value();
        std::string type = blocknode.attribute("type").value();

        if((name.length() == 0) || (type.length() == 0)){
            throw std::runtime_error("block name or type missing");
        }
        std::string threadpool=blocknode.attribute("threadpool").value();

        invocation_type invocation = invocation_type::Direct;
        
        /* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * 
         * OLD AND BUSTED sched_type attribute
         * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
        std::string sched_type=blocknode.attribute("sched_type").value();
        if (sched_type.length() > 0) {
            if (sched_type.compare("active") == 0) {
                invocation = invocation_type::Indirect;
            } else {
                std::runtime_error("sched_type val invalid");
            }
        }

        /* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * 
         * NEW invocation attribute (wins)
         * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
        std::string invocation_str=blocknode.attribute("invocation").value();
        if (invocation_str.length() > 0) {
            if (invocation_str.compare("direct") == 0) {
                invocation = invocation_type::Direct;
            } else if (invocation_str.compare("indirect") == 0) {
                invocation = invocation_type::Indirect;
            } else if (invocation_str.compare("async") == 0) {
                invocation = invocation_type::Async;
            } else {
                 std::runtime_error("invalid invocation type: use direct, indirect, async");
            }
        }


        create_block_from_parsed(type, name, invocation, blocknode);

        if (invocation == invocation_type::Indirect || invocation == invocation_type::Async) {
            if (threadpool.length() == 0) {
                throw std::runtime_error("need a threadpool for active/async invocation");
            }
            PoolManager::instance().add_to_pool(threadpool, m_map[name]);
        }
    }

    for (pugi::xml_node link=config.child("connection"); link; link=link.next_sibling("connection"))
    {
        std::string from_block=link.attribute("src_block").value();
        if (from_block.length() == 0) throw std::runtime_error("missing begin block in connection");
        
        std::string from_gate=link.attribute("src_gate").value();
        if (from_gate.length() == 0) throw std::runtime_error("missing begin gate in connection");
        
        std::string to_block=link.attribute("dst_block").value();
        if (to_block.length() == 0) throw std::runtime_error("missing dst block in connection");
        
        std::string to_gate=link.attribute("dst_gate").value();
        if (to_gate.length() == 0) throw std::runtime_error("missing begin gate in connection");
        
        create_link(from_block, from_gate, to_block, to_gate);
    }

    // initialize the configured blocks...
    //

    initialize();
}

File: daemon/core/composition.py
# Copyright (c) 2011, NEC Europe Ltd, Consorzio Nazionale 
# Interuniversitario per le Telecomunicazioni, Institut 
# Telecom/Telecom Bretagne, ETH Zuerich, INVEA-TECH a.s. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without 
# modification, are permitted provided that the following conditions are met:
#    * Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#    * Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in the
#      documentation and/or other materials provided with the distribution.
#    * Neither the names of NEC Europe Ltd, Consorzio Nazionale 
#      Interuniversitario per le Telecomunicazioni, Institut Telecom/Telecom 
#      Bretagne, ETH Zuerich, INVEA-TECH a.s. nor the names of its contributors 
#      may be used to endorse or promote products derived from this software 
#      without specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR 
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT 
# HOLDERBE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR 
# PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER 
# IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE
#

import os

from threadpool import ThreadPool
from block import Block, Connection
from blockinfo import block_infos
from core.returnvalue import *
from symbol import except_clause

class CompositionManager:

    """\brief Manages a blockmon composition
    """

    def __init__(self, comp_id, blockmon, bm_logger):
        """\brief Initializes class
        \param comp_id   (\c string)         The composition id
        \param blockmon  (\c module)         The blockmon executable module
        \param bm_logger (\c logging.logger) The logger
        """
        self.__logger = bm_logger
        self.__comp_id = str(comp_id)
        self.__blocks = {}         # dict block id -> Block
        self.__thread_pools = {}   # dict pool id  -> ThreadPool
        self.__blockmon = blockmon
        self.__blockmon.add_composition(self.__comp_id)

    def install(self, xmlnode):
        """\brief Installs a blockmon composition
        \param xmlnode (\c xml.dom.Node) The blockmon composition
        """
        
        generalXML = xmlnode.getElementsByTagName("general")
        if not generalXML:
            print "error, composition has no 'general' tag"
            os._exit(1)
        self.__process_general(generalXML[0])
            
        for n in xmlnode.getElementsByTagName("threadpool"):
            self.__create_pool(n)

        for n in xmlnode.getElementsByTagName("block"):
            self.__create_block(n)

        for n in xmlnode.getElementsByTagName("connection"):
            self.__create_connection(n)

        self.__blockmon.init_composition(self.__comp_id)
        
        self.__logger.info(self.__comp_id + ' : installed')
           
    def reconfigure(self, xmlnodes):
        """\brief Reconfigures a blockmon composition
        \param xmlnodes (\c xml.dom.Node) The blockmon composition
        """
        for xmlnode in xmlnodes.getElementsByTagName("delete"):
            for n in xmlnode.getElementsByTagName("connection"):
                self.__delete_connection(n)

            for n in xmlnode.getElementsByTagName("block"):
                self.__delete_block(n)

            for n in xmlnode.getElementsByTagName("threadpool"):
                self.__delete_pool(n)

        for xmlnode in xmlnodes.getElementsByTagName("add"):
            for n in xmlnode.getElementsByTagName("threadpool"):
                self.__create_pool(n)

            for n in xmlnode.getElementsByTagName("block"):
                self.__create_block(n)

            for n in xmlnode.getElementsByTagName("connection"):
                self.__create_connection(n)

        for xmlnode in xmlnodes.getElementsByTagName("reconf"):
            
            for n in xmlnode.getElementsByTagName("block"):
                self.__update_block(n)

        self.__logger.info(self.__comp_id + ' : updated')

    def remove(self):
        """\brief Removes the currently installed blockmon composition
        """
        for i in self.__blocks.values():
            i.remove()

        for t in self.__thread_pools.values():
            self.__blockmon.remove_thread_pool(t.name)

        self.__blockmon.delete_composition(self.__comp_id)

    def read_block_var(self, var_info):
        """\brief Reads a block variable's value
        \param var_info  (\c VariableInfo) The variable to read
        \return         (\c ReturnValue)  The result of the operation
        """
        blockname = var_info.get_block_name()
        varname = var_info.get_name()
        if not self.__blocks.has_key(blockname):
            return ReturnValue(ReturnValue.CODE_FAILURE, "block id " + blockname + " does not exist", None)
        block = self.__blocks[blockname]
#        if not block_infos.has_key(block.block_type):
#            return ReturnValue(ReturnValue.CODE_FAILURE, "no block of type "+block.block_type+" exists", None)
#        if not block_infos[block.block_type].has_variable(varname, "read"):
#            return ReturnValue(ReturnValue.CODE_FAILURE, "block has no read variable " + varname, None)
        try:
            value = block.read_var(varname)
        except RuntimeError, e:
            return ReturnValue(ReturnValue.CODE_FAILURE, "An exception occurred: '" + str(e)+"'", None)
        
        var_info.set_value(value)
        return ReturnValue(ReturnValue.CODE_SUCCESS, "", var_info)

    def write_block_var(self, var_info):
        """\brief Writes a value to a block variable
        \param var_info  (\c VariableInfo) The variable to write to
        \return         (\c ReturnValue)  The result of the operation
        """
        blockname = var_info.get_block_name()
        varname = var_info.get_name()
        value = var_info.get_value()
        if not self.__blocks.has_key(blockname):
            return ReturnValue(ReturnValue.CODE_FAILURE, "block id " + blockname + " does not exist", None)
        block = self.__blocks[blockname]
#        if not block_infos.has_key(block.block_type):
#            return ReturnValue(ReturnValue.CODE_FAILURE, "no block of that type exists", None)
#        if not block_infos[block.block_type].has_variable(varname, "write"):
#            return ReturnValue(ReturnValue.CODE_FAILURE, "block has no write variable " + varname, None)
        try:
            if self.__blocks[blockname].write_var(varname, value):
                return ReturnValue(ReturnValue.CODE_SUCCESS, "", None)
        except RuntimeError, e:
            return ReturnValue(ReturnValue.CODE_FAILURE, "An exception occurred: '" + str(e)+"'", None)
        
        return ReturnValue(ReturnValue.CODE_FAILURE, "error while writing to variable", None)

    def __process_general(self, xmlnode):
        """\brief Processes parameters under the 'general' tag
        \param xmlnode (\c xml.dom.Node) The XML for the general tag
        """
        clockXML = xmlnode.getElementsByTagName("clock")        
        clock_type = "WALL"
        if (clockXML and len(clockXML) > 0):
            clock_type = clockXML[0].attributes['type'].value
        self.__select_clock(str(clock_type))
        
    def __select_clock(self, clock_type):
        """\brief Sets the clock type (PACKET or WALL, case insensitive)
        \param clock_type (\c string)      The clock type
        \return           (\c ReturnValue) The result of the operation
        """
        clock_type = clock_type.upper()
        if clock_type == "WALL" or clock_type == "PACKET":
            self.__blockmon.select_clock(clock_type)
            return ReturnValue(ReturnValue.CODE_SUCCESS, "", None)
        return ReturnValue(ReturnValue.CODE_FAILURE, "invalid value for clock type", None)
    
    def __create_block(self, blocknode):
        """\brief Creates a block
        \param blocknode (\c xml.dom.Node) The block in xml form
        """
        block_name = blocknode.attributes['id'].value
        block_type = blocknode.attributes['type'].value
        
        invocation = "direct"
        tpool_id = ""
        tpool = None
                
#        params = blocknode.getElementsByTagName('params')
        #params = params[0].toxml()
        
        if blocknode.attributes.has_key('invocation'):
            invocation = blocknode.attributes['invocation'].value

        if invocation == "indirect" or invocation == "async":
            tpool_id = blocknode.attributes['threadpool'].value
            tpool = self.__thread_pools[str(tpool_id)]
        
        self.__blocks[block_name] = Block(block_name,\
                                          block_type,\
                                          self.__comp_id,\
                                          invocation,\
                                          blocknode.toxml(),\
                                          tpool_id,\
                                          tpool,\
                                          self.__blockmon,\
                                          self.__logger)

    def __delete_block(self, blocknode):
        """\brief Deletes a block
        \param blocknode (\c xml.dom.Node) The block in xml form
        """
        block_name = blocknode.attributes['id'].value
        self.__blocks[block_name].remove()
        del self.__blocks[block_name]

    def __update_block(self, blocknode):
        """\brief Updates a block
        \param blocknode (\c xml.dom.Node) The block in xml form
        """
        block_name = blocknode.attributes['id'].value
        
        params = blocknode.getElementsByTagName('params')
        params = str(params[0].toxml())
        
        invocation = "direct"
        tpool_id = ""
        tpool = None
        #active = False
 
        if blocknode.attributes.has_key('invocation'):
            invocation = blocknode.attributes['invocation'].value
        
        if invocation == "indirect" or invocation == "async":
            tpool_id = blocknode.attributes['threadpool'].value
            tpool = self.__thread_pools[str(tpool_id)]
            
        self.__blocks[block_name].update(invocation, params, tpool_id, tpool)
#        if blocknode.attributes.has_key('sched_type'):
#            active = True if (blocknode.attributes['sched_type'].value == 'active') else False
#            tpool_id = blocknode.attributes['threadpool'].value
#
#        if (active):
#            tpool = self.__thread_pools[str(tpool_id)]
#        self.__blocks[block_name].update(active, params, tpool_id, tpool)

    def __create_pool(self, poolnode):
        """\brief Creates a thread pool
        \param poolnode (\c xml.dom.Node) The thread pool in xml form
        """
        self.__thread_pools[poolnode.attributes['id'].value] = ThreadPool(poolnode,\
                                                                          self.__blockmon)
    def __delete_pool(self, poolnode):
        """\brief Deletes a thread pool
        \param poolnode (\c xml.dom.Node) The thread pool in xml form
        """
        del self.__thread_pools[poolnode.attributes['id'].value]

    def __create_connection(self, connection_node):
        """\brief Creates a connection between two blocks
        \param connection_node (\c xml.dom.Node) The connection in xml form
        """
        bsource = str(connection_node.attributes['src_block'].value)
        gsource = str(connection_node.attributes['src_gate'].value)
        bdst = str(connection_node.attributes['dst_block'].value)
        gdst = str(connection_node.attributes['dst_gate'].value)
        self.__connect_blocks(self.__blocks[bsource], gsource, self.__blocks[bdst], gdst)

    def __delete_connection(self, connection_node):
        """\brief Deletes a connection between two blocks
        \param connection_node (\c xml.dom.Node) The connection in xml form
        """
        bsource = str(connection_node.attributes['src_block'].value)
        gsource = str(connection_node.attributes['src_gate'].value)
        bdst = str(connection_node.attributes['dst_block'].value)
        gdst = str(connection_node.attributes['dst_gate'].value)
        self.__disconnect_blocks(self.__blocks[bsource], gsource, self.__blocks[bdst], gdst)

    def __connect_blocks(self, b1, g1, b2, g2):
        """\brief Connects two blocks
        \param b1  (\c Block)  The source block
        \param g1  (\c string) The source gate
        \param b2  (\c Block)  The destination block
        \param g2  (\c string) The destination gate
        """
        self.__blockmon.create_connection(b1.comp_id,\
                                          b1.name,\
                                          g1,\
                                          b2.comp_id,\
                                          b2.name,\
                                          g2)
        b1.add_connection(g1, b2, g2, 'out')
        b2.add_connection(g2, b1, g1, 'in')
        b1.connections.append(Connection(g1, b2, g2))
        b2.connections.append(Connection(g2, b1, g1))

    def __disconnect_blocks(self, b1, g1, b2, g2):
        """\brief Disconnects two blocks
        \param b1  (\c Block)  The source block
        \param g1  (\c string) The source gate
        \param b2  (\c Block)  The destination block
        \param g2  (\c string) The destination gate
        """
        self.__blockmon.delete_connection(b1.comp_id, \
                                          b1.name, \
                                          g1, \
                                          b2.comp_id, \
                                          b2.name, \
                                          g2)
        b1.remove_connection(g1, b2, g2, 'out')
        b2.remove_connection(g2, b1, g1, 'in')

    def get_blocks(self):
        return self.__blocks
